storage/backup/backup-cli/src/coordinators/restore.rs

```diff
@@ -88,72 +89,124 @@ impl RestoreCoordinator {
         )
         .await?;
 
-        let next_txn_version = self
+        let target_version = self.global_opt.target_version;
+        COORDINATOR_TARGET_VERSION.set(target_version as i64);
+
+        // calculate the start_version and replay_version
+        let max_txn_ver = metadata_view
+            .max_transaction_version()?
+            .ok_or_else(|| anyhow!("No transaction backup found."))?;
+        let lhs = self.ledger_history_start_version();
+
+        let latest_tree_version = self
+            .global_opt
+            .run_mode
+            .get_state_snapshot_before(Version::MAX);
+        let tree_completed = {
+            match latest_tree_version {
+                Some((ver, _)) => self
+                    .global_opt
+                    .run_mode
+                    .get_state_snapshot_before(ver)
+                    .is_some(),
+                None => false,
+            }
+        };
+
+        let mut db_next_version = self
             .global_opt
             .run_mode
             .get_next_expected_transaction_version()?;
-        if next_txn_version != 0 {
-            // DB is already in workable state
-            info!(
-                next_txn_version = next_txn_version,
-                "DB is ready to accept transactions, start the node to catch up with the chain. \
-                If the node is unable to catch up because the DB is too old, delete the data folder \
-                and bootstrap again.",
-            );
-            return Ok(());
-        }
 
-        let state_snapshot_backup =
-            if let Some(version) = self.global_opt.run_mode.get_in_progress_state_snapshot()? {
-                info!(
-                    version = version,
-                    "Found in progress state snapshot restore",
-                );
-                metadata_view.expect_state_snapshot(version)?
-            } else {
-                let max_txn_ver = metadata_view
-                    .max_transaction_version()?
-                    .ok_or_else(|| anyhow!("No transaction backup found."))?;
-                metadata_view
-                    .select_state_snapshot(std::cmp::min(self.target_version(), max_txn_ver))?
-                    .ok_or_else(|| anyhow!("No usable state snapshot."))?
-            };
-        let version = state_snapshot_backup.version;
-        self.global_opt.target_version = version;
-        let epoch_ending_backups = metadata_view.select_epoch_ending_backups(version)?;
-        let transaction_backups = metadata_view
-            .select_transaction_backups(self.ledger_history_start_version(), version)?;
-        COORDINATOR_TARGET_VERSION.set(version as i64);
-        info!(version = version, "Restore target decided.");
+        let kv_snapshot = match self.global_opt.run_mode.get_in_progress_state_kv_snapshot() {
+            Ok(Some(ver)) => {
+                // already restored the kv snapshot, no need to restore again
+                if db_next_version > ver {
```

Comment:
should be `>=`, right? when `db_next_version == ver`, does it mean the snapshot has done?